### ISSUE TITLE
MaskRCNN: Make return values more configurable

### ISSUE THREAD

**Issue by claushofmann** [pytorch/vision#6260]
### 🚀 The feature

Make MaskRCNN outputs configurable and independent of whether the model is in `train` or `eval` mode.

How it is currently done (according to the documentation):

> During training, the model expects both the input tensors, as well as a targets (list of dictionary),
    containing:
        - boxes (FloatTensor[N, 4]): the ground-truth boxes in [x1, y1, x2, y2] format, with values of x
          between 0 and W and values of y between 0 and H
        - labels (Int64Tensor[N]): the class label for each ground-truth box
        - masks (UInt8Tensor[N, H, W]): the segmentation binary masks for each instance

>   The model returns a Dict[Tensor] during training, containing the classification and regression
    losses for both the RPN and the R-CNN, and the mask loss.

 >  During inference, the model requires only the input tensors, and returns the post-processed
    predictions as a List[Dict[Tensor]], one for each input image. The fields of the Dict are as
    follows:
        - boxes (FloatTensor[N, 4]): the predicted boxes in [x1, y1, x2, y2] format, with values of x
          between 0 and W and values of y between 0 and H
        - labels (Int64Tensor[N]): the predicted labels for each image
        - scores (Tensor[N]): the scores or each prediction
        - masks (UInt8Tensor[N, 1, H, W]): the predicted masks for each instance, in 0-1 range. In order to
          obtain the final segmentation masks, the soft masks can be thresholded, generally
          with a value of 0.5 (mask >= 0.5)

How it would be nicer IMHO:

```

model = MaskRCNN(...)

model.train()

detections, losses = model(image, target)  # should output both detections and losses, not only losses

model.eval()

detections = model(image)
detections, losses = model(image, target)  # should output both detections and losses, not only detections
```

### Motivation, pitch

This would dramatically improve evaluation of the model. Currently, (at least from what I have seen) it is not easily possible to get the validation loss for the MaskRCNN. In theory, one could pass the validation set in training mode and then obtain the loss, but this could have unintended side-effects to other parts of the model (e.g. on dropout if there is one)

**Comment by NicolasHug** [comment]
Thanks for opining this issue @claushofmann 

This has been a requested feature for a while e.g. in https://github.com/pytorch/vision/issues/1574 or https://github.com/pytorch/vision/issues/4749

There are 2 main difficulties for us here:
- Torchscript used to not support `Union`, which made any change almost impossible. Now that it supports `Union`, things might get easier
- it's hard to make any change without introducing somewhat complex parameter, or without breaking the current API. Since all of our `detection` module is still in Beta stage (although this isn't really obvious for users), we might at some point go this route though.

I'll close this one to avoid duplication if you don't mind, as we keep track of this in https://github.com/pytorch/vision/issues/1574


CCing @datumbox as this is a fairly common request that it would be nice to address eventually
